### PR TITLE
ORC-751: [C++] Implement Predicate Pushdown for C++ Reader

### PR DESCRIPTION
**Pull request by wgtmac**
1. Use RowReaderOptions to pass SearchArgument to enable PPD.
2. Modify RowReaderImpl::startNextStripe to seek to next matched row group based on sarg evaluation.
3. RowReaderImpl::next seeks to next matched row group based on sarg evaluation.
4. RowReaderImpl::seekToRow also jumps to the 1st matched row group after specified row number.

### CODE REVIEW

**Comment by wgtmac** [comment]
@dongjoon-hyun I have closed the old PR and moved it here. Please review it again. Thanks!

File: c++/src/sargs/SargsApplier.cc [hunk 1]
```diff
@@ -25,7 +26,13 @@ namespace orc {
                                     const std::string& colName) {
     for (uint64_t i = 0; i != type.getSubtypeCount(); ++i) {
       if (type.getFieldName(i) == colName) {
-        return type.getSubtype(i)->getColumnId();
+        if (type.getKind() == CHAR || type.getKind() == VARCHAR) {
```
Comment (pgaref) [inline c++/src/sargs/SargsApplier.cc:29]: Shall we leave this to the [consumer](https://github.com/apache/orc/blob/295ea5be6c2ae643bb3c3ca6db7b018aeb51bebc/java/core/src/java/org/apache/orc/impl/SchemaEvolution.java#L378)? We dont have anything similar for the Java side.
We probably should address this as part of ORC-612

**Review by pgaref — COMMENTED** [review]
@wgtmac Changes look good to me in general left some comments.
Please let me know what you think!

File: c++/include/orc/Reader.hh [hunk 2]
```diff
@@ -538,6 +549,12 @@ namespace orc {
      */
     virtual void seekToRow(uint64_t rowNumber) = 0;
 
+    /**
+     * If PPD is enabled, returns true and store number of selected RGs and
+     * number of evaluated RGs into the stats pair; otherwise returns false.
+     */
+    virtual bool getPPDStats(std::pair<uint64_t, uint64_t>& stats) const = 0;
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply c++/include/orc/Reader.hh:556]: +1 for @pgaref 's comment.

File: c++/src/Reader.cc [hunk 3]
```diff
@@ -68,6 +68,12 @@ namespace orc {
       return columnPath.substr(0, columnPath.length() - 1);
   }
 
+  WriterVersion getWriterVersionImpl(const FileContents * contents) {
+    if (!contents->postscript->has_writerversion()) {
+      return WriterVersion_ORIGINAL;
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply c++/src/Reader.cc:73]: +1 for @pgaref 's comment.

File: c++/src/Reader.hh [hunk 4]
```diff
@@ -142,6 +143,30 @@ namespace orc {
 
     // row index of current stripe with column id as the key
     std::unordered_map<uint64_t, proto::RowIndex> rowIndexes;
+    std::map<uint32_t, BloomFilterIndex> bloomFilterIndex;
```
Comment (dongjoon-hyun) [inline c++/src/Reader.hh:146]: Shall we remove this in this PR because this seems to be unused in this PR?

File: c++/src/sargs/SargsApplier.cc [hunk 5]
```diff
@@ -25,7 +26,13 @@ namespace orc {
...
+          ///FIXME: disable PPD for char and varchar types as their rules
+          // vary among different engines.
+          return INVALID_COLUMN_ID;
```
Comment (dongjoon-hyun) [inline c++/src/sargs/SargsApplier.cc:32]: Do we have a test coverage for this code path?
If not, I'm +1 for @pgaref 's opinion. We can do this later.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with some minor comments)
Shall we address them and merge this?

File: c++/src/Reader.hh (hunk 4 above)
Comment (wgtmac) [reply c++/src/Reader.hh:146]: it is used to hold bloom filter index in Reader.cc (e.g. line 378)

File: c++/src/Reader.cc [hunk 6]
```diff
@@ -944,13 +1042,67 @@ namespace orc {
...
     }
     return rowsToRead != 0;
   }
 
+  uint64_t RowReaderImpl::computeBatchSize(uint64_t requestedSize,
+                                           uint64_t currentRowInStripe,
+                                           uint64_t rowsInCurrentStripe,
+                                           uint64_t rowIndexStride,
+                                           const std::vector<bool>& includedRowGroups) {
+    uint64_t endRowInStripe = rowsInCurrentStripe;
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/src/Reader.cc:1077]: Added a comment here. The logic is slightily different with the java side so I'd prefer to keep the logic here.

File: c++/src/Reader.cc [hunk 7]
```diff
@@ -892,29 +939,67 @@ namespace orc {
...
+        } else {
+          // advance to next stripe when current stripe has no matching rows
+          currentStripe += 1;
+          currentRowInStripe = 0;
+        }
+      }
+    } while (sargsApplier && currentStripe < lastStripe);
+
+    if (currentStripe < lastStripe) {
+      const Timezone& writerTimezone =
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/src/Reader.cc:981]: Done

File: c++/src/sargs/SargsApplier.cc (hunk 1 above)
Comment (wgtmac) [reply c++/src/sargs/SargsApplier.cc]: Sure

File: c++/include/orc/Reader.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgaref) [reply c++/include/orc/Reader.hh:556]: Agreed, lets remove this for now and use a separate ticket to track.
I believe the remaining changes are ready to be merged after that.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @wgtmac . 
Do you have other comments, @pgaref ?

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by pgaref** [comment]
> Thank you for updating, @wgtmac .
> Do you have other comments, @pgaref ?

Hey @dongjoon-hyun PR looks good to me as well, we just need to remove the newly added getPPDStats() API before merging.

**Comment by pgaref** [comment]
> > > Thank you for updating, @wgtmac .
> > > Do you have other comments, @pgaref ?
> > 
> > 
> > Hey @dongjoon-hyun PR looks good to me as well, we just need to remove the newly added getPPDStats() API before merging.
> 
> It has been removed now. Thanks for reviewing the PR! @pgaref @dongjoon-hyun

Thanks for updating @wgtmac ! seems like mStats can also be removed as its not longer used but its less important.
+1 from my side

**Comment by dongjoon-hyun** [comment]
Thank you, @wgtmac and @pgaref . Merged to master!

**Pull request merged**